sed/calibration/energy.py

```diff
@@ -0,0 +1,112 @@
+import copy
+import gc
+from pathlib import Path
+from typing import Union
+
+import numpy as np
+import pandas as pd
+
+
+class EnergyCalibrator:
+    def __init__(
+        self,
+        df: pd.DataFrame = None,
+        tof_axis_name: str = "dldTime",
+        tof_axis_units: str = "step",
+        sector_correction=True,
+        jitter=True,
+        functional_shift=False,
+        parameters: Union[dict, str, Path] = {},
+        **kwargs,
+    ):
+        self._df = df
+        self.tof_axis_name = tof_axis_name
+        self.tof_axis_units = tof_axis_units
+        self._parameters = self.load_parameters(parameters, **kwargs)
+        self._cached_parameters = {}
+
+    def load_parameters(self, pars, **kwargs):
+        """Load parameters from file, dictionary or explicit arguments
+
+        Args:
+            pars: dictionary or file containing the required parameters
+            kwargs: any other argument passed is added as a parameter
+
+        Raises:
+            AttributeError: _description_
+
+        Returns:
+            the merged dictionary containing all parameters
+        """
+        if isinstance(pars, (str, Path)):
+            pars = self._parameters_from_file(pars)
+        elif not isinstance(pars, dict):
+            raise AttributeError(
+                "paramters must be a dictionary or a json/yaml file.",
+            )
+        return {**pars, **kwargs}
+
+    def _init_calibration_dataframe(self):
+        self._calib_df = copy.deepcopy(self._df[self.tof_axis_name])
+        if self.tof_axis_units == "step":
+            self._calib_df["tof_step"] = self._calib_df[self.tof_axis_name]
+            self._calib_df["tof_ns"] = self._calib_df["tof_step"]
+
+    @property
+    def parameters(self):
+        return {**self._parameters, **self._cached_parameters}
+
+    def clear_cache(self):
+        """clear the cached datasets and collect garbage"""
+        del self._cached_parameters
```

Comment:
what are these cached_parameters good for? are the filled anywhere?

Reply:
As above, these are not implemented yet. The idea here is to handle long computations of averages and cache them for future use, when testing parameters.